### PR TITLE
Adding implementations for:

### LINKED ISSUE #710: Missing common functions find in DataEngineering
## What is the current behavior?
The following functions:

function.array_distinct
functions.regexp_extract
functions.struct
functions.format_number
function.daydiff
functions.date_add
functions.date_sub

Are not available.

## What is the desired behavior?
Provide equivalent for these functions as they are present in many existing DE workloads

## How would this improve `snowflake-snowpark-python`?
It will make it easier to move existing workloads to snowpark, as less changes will be needed.

### PR DESCRIPTION
**Pull request by sfc-gh-mrojas**
function.array_distinct
functions.regexp_extract
functions.struct
functions.format_number
function.daydiff
functions.date_add
functions.date_sub

Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #710

2. Fill out the following pre-review checklist:

   - [X ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.
It provides implementations for the aforementioned functions

### CODE REVIEW

**Comment by sfc-gh-mrojas** [comment]
I have read the CLA Document and I hereby sign the CLA

File: src/snowflake/snowpark/functions.py [hunk 1]
```diff
@@ -1371,6 +1386,29 @@ def upper(e: ColumnOrName) -> Column:
     c = _to_col_if_str(e, "upper")
     return builtin("upper")(c)
 
+def struct(*cols)    -> Column:
+    """Creates a new object column.
+    Example::
+        >>> df = session.create_dataframe(['a', 'b', 'c'], schema=["a", "b", "c"])
+        >>> df.select(struct("a", "b", "c")).collect()
+    """
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/functions.py:1394]:
```suggestion
def struct(*cols) -> Column:
    """Creates a new object column.
    
    Example::
        >>> df = session.create_dataframe(['a', 'b', 'c'], schema=["a", "b", "c"])
        >>> df.select(struct("a", "b", "c")).collect()
    """
```

please also add output that come from `.collect()`

File: src/snowflake/snowpark/functions.py [hunk 2]
```diff
@@ -1474,6 +1512,17 @@ def regexp_count(
     params = [lit(p) for p in parameters]
     return builtin(sql_func_name)(sub, pat, pos, *params)
 
+def regexp_extract(value:ColumnOrLiteralStr,regexp:ColumnOrLiteralStr,idx:int) -> Column:
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/functions.py:1515]: This needs linting too

**Review by sfc-gh-aalam — COMMENTED** [review]
I don't understand why we are merging to `snowflakedb:main` instead of `main` branch

File: src/snowflake/snowpark/functions.py (hunk 1 above)
Comment (sfc-gh-mrojas) [reply src/snowflake/snowpark/functions.py]: output added

File: src/snowflake/snowpark/functions.py (hunk 2 above)
Comment (sfc-gh-mrojas) [reply src/snowflake/snowpark/functions.py]: lint aplied in last commit

**Pull request closed without merging**